### PR TITLE
Make bandwidth routing use routingDecision channel.

### PR DESCRIPTION
**Pull request by djpetti**
This method of communicating is more standard than what was in
use before. I also rethought some of the dependency injection
stuff to make initialization code cleaner.

---

This change is [Reviewable](https://reviewable.io/reviews/autoroute/node/78)

### CODE REVIEW

File: internal/bandwidth.go [hunk 1]
```diff
@@ -2,56 +2,120 @@ package internal
...
+
 	// Total number of bytes sent.
 	bytes_sent map[types.NodeAddress]int64
 	// Total elapsed time waiting for that data to send.
 	sent_time map[types.NodeAddress]time.Duration
 	// Time at which the last packet send started.
 	send_start_time map[types.NodeAddress]time.Time
+	// This is so we can keep track of packets that came from the same source. It
+	// maps each source to the next hop of the last packet sent from that source.
+	last_sent_to map[types.NodeAddress]types.NodeAddress
```
Comment (c00w) [inline internal/bandwidth.go:31]: This might be better expressed as inbound + outbound bandwidth.

File: internal/bandwidth.go [hunk 2]
```diff
@@ -63,9 +127,11 @@ func (b *bandwidthEstimator) SentPacket(node types.NodeAddress, size int64) {
 func (b *bandwidthEstimator) GetWeights(nodes []types.NodeAddress) []float64 {
 	// First, we need to calculate the total bandwidth so we know how to scale it.
 	total := float64(0)
+	b.bandwidth_lock.RLock()
```
Comment (c00w) [inline internal/bandwidth.go:130]: It might be easier to just grab the lock once and defer on the unlock.

File: internal/bandwidth_routing.go [hunk 3]
```diff
@@ -50,45 +47,35 @@ func chooseNextHop(weights []float64,
...
-// Send a packet. See routingAlgorithm interface for details.
-func (b *bandwidthRouting) SendPacket(dest DataConnection,
-                                      addr types.NodeAddress,
-                                      packet types.Packet) error {
-	b.bandwidth_mutex.Lock()
-	b.bandwidths.WillSendPacket(addr)
-	b.bandwidth_mutex.Unlock()
+// Sets the routing handler that we will use with this algorithm.
+// See the routingAlgorithm interface for details.
+func (b *bandwidthRouting) BindToRouting(routing *routingHandler) {
```
Comment (c00w) [inline internal/bandwidth_routing.go:73]: Can you just pass in the bandwidth estimator at construction time / always create it?

**Comment by c00w** [comment]
LGTM, feel free to merge, minor comments around some structure stuff.

**Comment by djpetti** [comment]
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

_[internal/bandwidth.go, line 31 [r1]](https://reviewable.io:443/reviews/autoroute/node/78#-KF75eAyggK28vKqEZAg-r1-31:-KFGZBhJ9LC8Bt4lDGxz:-125444377) ([raw file](https://github.com/autoroute/node/blob/08e1ae329f7c72d00b3920479745ba6118ae2718/internal/bandwidth.go#L31)):_
Yeah, that's not quite what this is doing, though. It's there to make sure that when we send one packet to one destination and then the next packet to another destination, it has enough sense to conclude that the first send completed. Without this, it doesn't quite have enough information to make that deduction.

---

_[internal/bandwidth.go, line 130 [r1]](https://reviewable.io:443/reviews/autoroute/node/78#-KF75eAyggK28vKqEZAg-r1-130:-KFGYR2UZZV8tqwoGMbk:-254713290) ([raw file](https://github.com/autoroute/node/blob/08e1ae329f7c72d00b3920479745ba6118ae2718/internal/bandwidth.go#L130)):_
Fair enough. Normally I would do that, but I assumed go used futex calls under the hood, which are almost free. In this case, though, I think your way makes for cleaner code, and I'm obsessing about negligible performance improvements.

---

_[internal/bandwidth_routing.go, line 73 [r1]](https://reviewable.io:443/reviews/autoroute/node/78#-KF75eAyggK28vKqEZAh-r1-73:-KFG_ClF7h5TW3FlKy-P:1839833500) ([raw file](https://github.com/autoroute/node/blob/08e1ae329f7c72d00b3920479745ba6118ae2718/internal/bandwidth_routing.go#L73)):_
There's actually some tricky dependency stuff involved here. Mainly, bandwidthEstimator needs access to the Routes channel of a routingHandler. Since certain routingAlgorithms need access to a bandwidthEstimator object, that means that we need to somehow get a routingHandler to a routingAlgorithm. This is hard, because right now it is set up so that there is a single routingHandler and the routingAlgorithm it uses can be injected at construction time. I really like that paradigm, because it makes switching algorithms trivial and obvious. I therefore judged that this one annoying method was a reasonable price to pay for it.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/autoroute/node/78)_

**Pull request merged**